### ISSUE TITLE
Support for config files

### ISSUE THREAD

**Issue by ebattenberg** [benanne/Lasagne#111]
I've been building enhancements on top of @benanne's Galaxy Challenge code for a few months for my work.  I think I'm ready to transition over to contributing to Lasagne.  One of the things I added to my (private) fork is support for constructing models and configuring and running training using YAML files.  I like this approach because it decouples the hyperparameters from the Python boilerplate and makes iterating over hyperparameters a lot cleaner.  I'd be happy to work on this functionality if it's within the scope and goals of the project.  How do others feel about this?

**Comment by benanne** [comment]
Cool :) To be honest I've never been a fan of configuration files / languages, I just keep my architecture definitions (and other hyperparameters) in separate Python files these days. This affords me maximal flexibility if I want to do something special. (For the galaxy challenge I didn't do that, mainly out of laziness).

That said, I'm not opposed to it, as long as it doesn't add a considerable burden for us to maintain compatibility with this format. Creating extra dependencies in the project that we have to keep in mind at all times is the main thing I'm worried about.

Personally if I had to build a config format I would opt for JSON, since it's a lot more human-readable and easier to learn. But I appreciate that YAML could be a lot more flexible. As long as it's not XML, I guess!

**Comment by ebattenberg** [comment]
This is the format I've been using.  I chose YAML because I want it to be as human readable and easy to edit by hand as possible (unlike protocol buffers).  

https://gist.github.com/ebattenberg/f5ea4b958edc3869515d#file-gistfile1-yml

And I personally think YAML is a lot more readable and easier to edit than JSON.  Though I agree the JSON is more universal.

**Comment by ebattenberg** [comment]
For comparison, here's the same converted to JSON (I cleaned it up a little and used the indent option)

https://gist.github.com/ebattenberg/6ea175765c61ffeb7952

**Comment by benanne** [comment]
That's interesting, your YAML definitely looks a lot less alien than what I'm used to seeing. Most YAML config files seem to make copious use of syntax that's unintuitive (!, &, ...) and I can't tell what it means just by looking at it. I could look it up and learn the syntax, but I'd rather not :)

If we can stick to the "intuitive" subset of YAML it does look a lot cleaner and easier to read.

One possible problem with this format is that it only seems to support stacks of layers. How would you handle trees or arbitrary DAGs?

Also I noticed some of the identifiers do not match the names used in the library (`rectifier` instead of `rectify` for example) and I think it would be important to be consistent there.

**Comment by f0k** [comment]
DAGs are easy, instead of naming your layers `0`, `1`, `2` as in Eric's example, you can give them arbitrary names and refer to the input layer(s) by their name(s).
Still I'm not quite sure if that should become part of Lasagne. If at all, it should probably be limited to describing the network architecture, not the training parameters.
In the beginning I was very fond of designing and implementing a YAML architecture description as well, but I've given in to using separate Python files for architecture, dataset and training description (dynamically loaded into the main driver script via the `imp` module) and I don't regret it -- it really allows you a lot more flexibility (such as whipping up a `Layer` class right within your architecture definition, or using a loop to create a configurable number of stacked layers, or whatever you can express in Python).

**Comment by benanne** [comment]
> (dynamically loaded into the main driver script via the `imp` module) 
ha, interesting. I'm using `importlib`. I wonder why both of these exist, doesn't the Zen of Python say "There should be one-- and preferably only one --obvious way to do it."?

**Comment by ebattenberg** [comment]
>That's interesting, your YAML definitely looks a lot less alien than what I'm used to seeing. Most YAML config files seem to make copious use of syntax that's unintuitive (!, &, ...) and I can't tell what it means just by looking at it. I could look it up and learn the syntax, but I'd rather not :) and I think a lot of people would be just as lazy.

Yeah, I thought that about YAML at first too.  YAML can be really simple.  I think all those confusing syntax additions are for communicating complicated stuff to the target language (like which classes to instantiate).  I like to use YAML just for its clean dictionary/hash type.  If code that uses the config file know how to convert a string into a class (via a dictionary mapping strings to classes), then the config can stay simple.

>One possible problem with this format is that it only seems to support stacks of layers. How would you handle trees or arbitrary DAGs?

I've thought about that a bit but haven't needed that functionality yet, so haven't tackled it.  A couple of ideas are:
1. Nesting parallel compound layers in a list, for example, if two layers take the same input and produce parallel output that is then fed into a concatenate layer.  
2. Having input/output index arguments to the layer entries.  Oh, what Jan said above... :)


For Jan's response:
>DAGs are easy, instead of naming your layers 0, 1, 2 as in Eric's example, you can give them arbitrary names and refer to the input layer(s) by their name(s).
Still I'm not quite sure if that should become part of Lasagne. If at all, it should probably be limited to describing the network architecture, not the training parameters.
In the beginning I was very fond of designing and implementing a YAML architecture description as well, but I've given in to using separate Python files for architecture, dataset and training description (dynamically loaded into the main driver script via the imp module) and I don't regret it -- it really allows you a lot more flexibility (such as whipping up a Layer class right within your architecture definition, or using a loop to create a configurable number of stacked layers, or whatever you can express in Python).

I agree that using Python scripts for configuration is a lot more flexible.  The YAML config option would definitely include only the most common subset of architectures and options and mainly be an interface for users who want to configure commonly used model types.

**Comment by dnouri** [comment]
IMO, cuda-convnet still has the best configuration format as far as
neural net libraries go.  I like the INI-style a lot more than any other
format:
https://code.google.com/p/cuda-convnet2/source/browse/layers/layer-params-cifar10-11pct.cfg
https://code.google.com/p/cuda-convnet2/source/browse/layers/layers-cifar10-11pct.cfg

But I too prefer Python over any configuration format, because
flexibility, and no "gap".  Though there's this advantage that those INI
files and other configuration formats have, and it's that you have all
the parameters in one place.  I'd argue that if we had a somewhat
standardized, declarative way of defining models (the Model class?),
then that would already fix most of the reasons why people would want to
use a configuration file.

**Comment by ebattenberg** [comment]
It sounds like everyone has their own favorite way to deal with this. ;)  I'll hold off on a PR.

Even though writing straight Python scripts with the configuration is more flexible, I'd still argue that separating data and parameters from code is in general a good idea.  So even if there isn't official support for a specific config file format in Lasagne, the hyperparameters you use in your custom scripts should still be coming from some sort of config file, whether it's json, yaml, ini, or protobufs.  It's just so much easier to edit and see at once than when your hard-coded constants are spread out all over your code.

**Comment by benanne** [comment]
I agree that separating parameters from code is good, but I feel that this separation isn't always so clear cut w.r.t. what's "parameters" and what's "code" for a specific project or task, and shouldn't necessarily be fixed on beforehand. Parameters don't necessarily have to be number or strings: In many cases, the entire architecture is "parameters", i.e. not just the numbers of units, but also how they are connected. In that case especially, "code as configuration" is really useful.

I feel that the extra layer of abstraction of a "configuration system" just gets in the way, because if it can't do something you want to do, you will have to think about extending it, which distracts you from the main task at hand. And it's very hard to anticipate all the things you're going to want to do, at least for me :)

If configuration is code, the responsibility of separating the code and "parameters" is then your own - for example, you could throw in a quick layer definition as @f0k mentioned, but in the long term it's better to factor that out into its own module.

You have to maintain the separation yourself, it hasn't been fixed on beforehand. But if you're disciplined enough to actually do this, I think the pros outweigh the cons by far. You could even shift the line of separation very easily while the work is in progress, and move some stuff into your config files that wasn't configurable previously.

Incidentally, one thing I really liked about Ruby on Rails and Django back in the day is that they also follow the "code as configuration" approach. I just think it's clean, with minimal cognitive overhead and maximal flexibility. But as before, with that flexibility comes the responsibility to maintain a clean separation yourself.

If you have a fixed network structure and you're just trying to find the best values for the learning rate, regularization, numbers of units etc., I agree that it's not necessary. But that's rarely how things work out for me.

**Comment by craffel** [comment]
I think this is potentially a very useful thing, but basically because
 
> It sounds like everyone has their own favorite way to deal with this.

I'd suggest that it be kept out of Lasagne.  I think keeping Lasagne's scope as convenient building blocks for arbitrarily structured nets is a good idea, and if people want to develop other libraries (e.g. a YAML configuration utility) on top of it then that's great.  Just my 2c.

**Comment by dimatura** [comment]
I've tried dozens of different ways of configuring my code and I've never been really happy with any solution. Definitely a hard thing to get right. 

Personally I'm using the YAML system from `pylearn2`. It's not pretty (certainly less clean that the one from @ebattenberg ) but it works. Definitely considering moving to just using python modules. 

Either way, one thing I've missed that would make configuration and post-hoc analysis more intuitive is a 'name' member for layers. Has anyone else felt this need?

**Comment by dnouri** [comment]
On 01/25/2015 07:27 PM, Daniel Maturana wrote:
> Either way, one thing I've missed that would make configuration and
> post-hoc analysis more intuitive is a 'name' member for layers. Has
> anyone else felt this need?

Yes, I've felt it:

https://github.com/dnouri/kfkd-tutorial/blob/master/kfkd.py#L180-L193

**Comment by benanne** [comment]
Maybe someone can make a new issue for this, so we don't go off topic here.

**Comment by ebattenberg** [comment]
Thanks for your input @craffel.  I think holding off on this is probably the best choice right now.  I'll reopen this if I come up with anything worth updating about.
